operator/watchers/service_sync.go

```diff
@@ -0,0 +1,264 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"context"
+	"log/slog"
+	"maps"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+
+	operatorOption "github.com/cilium/cilium/operator/option"
+	"github.com/cilium/cilium/pkg/annotation"
+	serviceStore "github.com/cilium/cilium/pkg/clustermesh/store"
+	cmtypes "github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/k8s"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/kvstore/store"
+	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/promise"
+)
+
+type ServiceSyncCallback func(context.Context)
+
+// ServiceSyncCell implements synchronization of Kubernetes services and endpoints
+// to kvstore.
+var ServiceSyncCell = cell.Module(
+	"service-sync",
+	"Synchronizes Kubernetes services to KVStore",
+
+	cell.Invoke(registerServiceSync),
+	cell.Provide(newClusterServiceConverter),
+)
+
+type ClusterServiceConverter interface {
+	Convert(svc *slim_corev1.Service, eps []*k8s.Endpoints) (out *serviceStore.ClusterService, toUpsert bool)
+	ForDeletion(svc *slim_corev1.Service) (out *serviceStore.ClusterService)
+}
+
+type ServiceSyncParams struct {
+	cell.In
+
+	Log                     *slog.Logger
+	ClusterInfo             cmtypes.ClusterInfo
+	Clientset               k8sClient.Clientset
+	Services                resource.Resource[*slim_corev1.Service]
+	Endpoints               resource.Resource[*k8s.Endpoints]
+	StoreFactory            store.Factory
+	ClusterServiceConverter ClusterServiceConverter
+	Backend                 promise.Promise[kvstore.BackendOperations] `optional:"true"`
```

Comment:
Little bit messy doing it this way. These are not provided by operator's hive, but are provided in clustermesh-apiserver. Should probably look into switching the operator to using `kvstore.Cell`

Reply:
Finally managed to tackle that refactoring: https://github.com/cilium/cilium/pull/39927